File: ruby/triangle/triangle.rb
class Triangle
  private

  attr_reader :a, :b, :c, :valid

  public

  def initialize(sides)
    @a, @b, @c = sides.sort
    @valid = @a.positive? && (@a + @b - @c).positive?
  end

  def equilateral?
    valid && a == b && b == c
  end

  def isosceles?
    valid && (a == b || a == c || b == c)
  end

  def scalene?
    valid && !equilateral? && !isosceles?
  end
end

File: clojure/triangle/src/triangle.clj
(ns triangle)

(defn- is-valid? [sides]
  (let [[a b c] (sort sides)]
    (and (pos? a)
         (< c (+ a b)))))

(defn- triangle-unique-sides-test [cmp & sides]
  (and (is-valid? sides)
       (-> sides set count cmp)))

(def equilateral? (partial triangle-unique-sides-test #(=  % 1)))
(def isosceles?   (partial triangle-unique-sides-test #(<= % 2)))
(def scalene?     (partial triangle-unique-sides-test #(=  % 3)))
